src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs

```diff
@@ -54,13 +49,64 @@ public void RegisterExecutionPipeline(Action<IncrementalGeneratorPipelineContext
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="source">The source code to add to the compilation</param>
+        public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8));
+
+        /// <summary>
+        /// Adds a <see cref="SourceText"/> to the compilation that will be available during subsequent phases
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="sourceText">The <see cref="SourceText"/> to add to the compilation</param>
+        public void AddSource(string hintName, SourceText sourceText) => AdditionalSources.Add(hintName, sourceText);
+    }
+
     public readonly struct IncrementalGeneratorPipelineContext
     {
-        public IncrementalValueSources Sources { get; }
+        private readonly ArrayBuilder<ISyntaxInputNode> _syntaxInputBuilder;
+        private readonly ArrayBuilder<IIncrementalGeneratorOutputNode> _outputNodes;
+
+        internal IncrementalGeneratorPipelineContext(ArrayBuilder<ISyntaxInputNode> syntaxInputBuilder, ArrayBuilder<IIncrementalGeneratorOutputNode> outputNodes)
+        {
+            _syntaxInputBuilder = syntaxInputBuilder;
+            _outputNodes = outputNodes;
+        }
+
+        public SyntaxValueProvider SyntaxProvider => new SyntaxValueProvider(_syntaxInputBuilder, RegisterOutput);
+
+        public IncrementalValueProvider<Compilation> CompilationProvider => new IncrementalValueProvider<Compilation>(SharedInputNodes.Compilation.WithRegisterOutput(RegisterOutput));
+
+        public IncrementalValueProvider<ParseOptions> ParseOptionsProvider => new IncrementalValueProvider<ParseOptions>(SharedInputNodes.ParseOptions.WithRegisterOutput(RegisterOutput));
+
+        public IncrementalValuesProvider<AdditionalText> AdditionalTextsProvider => new IncrementalValuesProvider<AdditionalText>(SharedInputNodes.AdditionalTexts.WithRegisterOutput(RegisterOutput));
+
+        public IncrementalValueProvider<AnalyzerConfigOptionsProvider> AnalyzerConfigOptionsProvider => new IncrementalValueProvider<AnalyzerConfigOptionsProvider>(SharedInputNodes.AnalyzerConfigOptions.WithRegisterOutput(RegisterOutput));
+
+        public void RegisterSourceOutput<TSource>(IncrementalValueProvider<TSource> source, Action<SourceProductionContext, TSource> action) => source.Node.RegisterOutput(new SourceOutputNode<TSource>(source.Node, action.WrapUserAction()));
+
+        public void RegisterSourceOutput<TSource>(IncrementalValuesProvider<TSource> source, Action<SourceProductionContext, TSource> action) => source.Node.RegisterOutput(new SourceOutputNode<TSource>(source.Node, action.WrapUserAction()));
 
```

Comment:
These methods do not access any instance state. Is that expected?

Reply:
Yeah, its by design, but a little unobvious. Previously they just called `_outputNodes.Add(...)`, then we moved them to the 'end of the chain' and they had to invoke the passed through `RegisterOutput` callback which would chain back to the top.

While doing that work I found a bug in the syntax nodes that means we wouldn't drop them even if unused, leading to lots of extra work. The callback chaining ensures we know when they are actually used. So even though we're moving the original calls back here, we still want to actually route through the callbacks. (But we want it to be an instance so it 'feels' like the user is registering it directly with the context).

We can probably refactor a lot of this, but I wanted to keep the code changes minimal while getting the public API shape correct.